### PR TITLE
#286: Fixed distributed writing to local filesystem

### PR DESCRIPTION
**Pull request by fhueske**
This fix enables local filesystem writes in a cluster setup.
Output directories for local FS writes are created on each TM but not on the JM (except if it hosts a TM as well).

The currently implemented override policy for local filesystem writes is:
- existing files is not deleted to create output directories (to avoid deleting OS files)
- existing files in existing output directories are overwritten

Let me know, if you want to enable deleting local FS files to create output directories.

### CODE REVIEW

File: stratosphere-runtime/src/main/java/eu/stratosphere/pact/runtime/task/DataSinkTask.java [hunk 1]
```diff
@@ -385,43 +385,52 @@ public int getMaximumNumberOfSubtasks()
...
-			try {
-				final FileStatus f = fs.getFileStatus(path);
-				if (f == null) {
-					return 1;
-				}
-				// If the path points to a directory we allow an infinity number of subtasks
-				if (f.isDir()) {
-					return -1;
+			
+			if(fs.getUri().getScheme().toLowerCase().equals("hdfs") || fs.getUri().getScheme().toLowerCase().equals("s3")) {
```
Comment (rmetzger) [inline stratosphere-runtime/src/main/java/eu/stratosphere/pact/runtime/task/DataSinkTask.java]: I'm not sure if this is the correct place for this.
I thought that file systems are pluggable. This here looks like hardcoded rules for specific file systems.

is there a way to move this code into the DistributedFileSystem (and S3) implementation?

The `FileOutputFormat` also contains specific `file:///` scheme code. I'm think In/Outputformats are independent of the filesystems (e.g. the file systems provide a data stream and the IF/OF converts it into an internal representation?)

Comment (fhueske) [inline stratosphere-runtime/src/main/java/eu/stratosphere/pact/runtime/task/DataSinkTask.java]: You are certainly right, that this is not the optimal place. In fact, most of that was already present and I only added a guard for distributed FS at this place. I was looking for a quick fix that doesn't break anything shortly before the release and didn't want to change too many things.

Now with the release out of the way, I can look for a proper solution based on file systems. However, the distinction between distributed and local file system will most likely remain in the DataSinkTask and the FileInputFormat as the distributed FS should be initialized at planning time (within the JM) and local FS need to be initialized in parallel at runtime...

Comment (rmetzger) [inline stratosphere-runtime/src/main/java/eu/stratosphere/pact/runtime/task/DataSinkTask.java]: Okay. I also think that it is a better idea to implement it correctly (e.g. no quick fix).

For the initialization at planning time: Can't you just add a new method to the FileSystems class for that?

**Comment by fhueske** [comment]
New PR for this: #421

**Pull request closed without merging**